sync-core/src/main/java/com/cloudant/sync/query/Index.java

```diff
@@ -0,0 +1,165 @@
+            logger.log(Level.SEVERE, "No index name was provided.");
+            return null;
+        }
+
+        if (indexType == null || indexType.isEmpty()) {
+            logger.log(Level.SEVERE, "No index type was provided.");
+            return null;
+        }
+
+        if (!validTypes.contains(indexType.toLowerCase())) {
+            logger.log(Level.SEVERE, String.format("Invalid index type %s.", indexType));
+            return null;
+        }
+
+        if (indexType.equalsIgnoreCase(JSON_TYPE) && indexSettings != null) {
+            logger.log(Level.WARNING, String.format("Index type is %s, index settings %s ignored.",
+                                                    indexType,
+                                                    indexSettings.toString()));
+            indexSettings = null;
+        } else if (indexType.equalsIgnoreCase(TEXT_TYPE)) {
+            if (indexSettings == null) {
+                indexSettings = new HashMap<String, String>();
+                indexSettings.put(TOKENIZE, DEFAULT_TOKENIZER);
+                logger.log(Level.INFO, String.format("Index type is %s, defaulting settings to %s.",
+                        indexType,
+                        indexSettings.toString()));
+            } else {
+                for (String parameter : indexSettings.keySet()) {
+                    if (!validSettings.contains(parameter.toLowerCase())) {
+                        String msg = String.format("Invalid parameter %s in index settings %s.",
+                                                   parameter,
+                                                   indexSettings);
+                        logger.log(Level.SEVERE, msg);
+                        return null;
+                    }
+                }
+            }
+        }
+
+        return new Index(fieldNames, indexName, indexType, indexSettings);
+    }
+
+    /**
+     * Compares the index type and accompanying settings with the passed in arguments.
+     *
+     * @param indexType the index type to compare to
+     * @param indexSettings the indexSettings to compare to
+     * @return true/false - whether there is a match
+     */
+    protected boolean compareIndexTypeTo(String indexType, String indexSettings) {
+        if (!this.indexType.equalsIgnoreCase(indexType)) {
+            return false;
+        }
+
+        if (this.indexSettings == null && indexSettings == null) {
+            return true;
+        } else if (this.indexSettings == null || indexSettings == null) {
+            return false;
+        }
+
+        String settingsAsAString;
+        try {
+            settingsAsAString = new ObjectMapper().writeValueAsString(this.indexSettings);
+        } catch (JsonProcessingException e) {
+            String msg = String.format("Error processing index settings %s",
+                                       this.indexSettings.toString());
+            logger.log(Level.SEVERE, msg, e);
+            return false;
+        }
```

Comment:
Your problem here might be in ordering semantics, the ordering of keys in the JSON from your indexSettings map might be different from the string passed in, even though the structures are semantically the same. I.e.,:

`{ "name: "mike", "age": 12}` vs. `{"age": 12, "name": "mike"}`.